lib/consensus/raft/consensus.go

```diff
@@ -0,0 +1,563 @@
+// Package raft implements a Consensus component for IPFS Cluster which uses
+// Raft (go-libp2p-raft).
+package consensus
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/google/uuid"
+
+	addr "github.com/filecoin-project/go-address"
+
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/node/config"
+
+	//ds "github.com/ipfs/go-datastore"
+	logging "github.com/ipfs/go-log/v2"
+	consensus "github.com/libp2p/go-libp2p-consensus"
+	rpc "github.com/libp2p/go-libp2p-gorpc"
+	libp2praft "github.com/libp2p/go-libp2p-raft"
+	host "github.com/libp2p/go-libp2p/core/host"
+	peer "github.com/libp2p/go-libp2p/core/peer"
+)
+
+var logger = logging.Logger("raft")
+
+type RaftState struct {
+	NonceMap map[addr.Address]uint64
+	MsgUuids map[uuid.UUID]*types.SignedMessage
```

Comment:
Did we decide if we want to use persistent raft state? If yes this will need to live in the datastore - possibly using https://github.com/filecoin-project/go-statestore (as an aside, we probably can make statestore a lot nicer to use with generics)

Reply:
Raft stores the serialization of the state in the snapshot store regularly, and the last index applied to it. So as long as you use a non-in-memory store (i.e Bolt), then you're good.

It is recommended that the RaftState implements Marshal/Unmarshal: https://github.com/libp2p/go-libp2p-raft/blob/master/consensus.go#L20-L36, although I'm guessing in this case the Msgpack default encoding should work.